### PR TITLE
Potential test fix and infra tests

### PR DESCRIPTION
**Pull request by bgavrilMS**

### CODE REVIEW

File: tests/Microsoft.Identity.Test.Integration.net45/Infrastructure/SeleniumWebUI.cs [hunk 1]
```diff
@@ -107,40 +107,72 @@ private IWebDriver InitDriverAndGoToUrl(string url)
...
+                    },
+                    cancellationToken);
+
+                // Run the TCP listener and the selenium automation in parallel
+                var seleniumAutomationTask = Task.Factory.StartNew(() => _seleniumAutomationLogic(driver));
 
-                            return GetMessageToShowInBroswerAfterAuth(uri);
-                        },
-                        cancellationToken);
+                await Task.WhenAny(seleniumAutomationTask, listenForAuthCodeTask)
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.net45/Infrastructure/SeleniumWebUI.cs]: I changed this logic from Task.WhenAll to Task.WhenAny. WhenAll actually waits for both tasks to finish, even if one fails fast. So any Selenium failure would still wait for the TCP listener to timeout, which takes 1 min. I assume this is causing VSTS to fail. 

When using WhenAny, the await will not throw! So we need to do manual processing of exceptions.

File: tests/Microsoft.Identity.Test.Integration.net45/Infrastructure/SingleMessageTcpListener.cs [hunk 2]
```diff
@@ -67,15 +62,27 @@ await AcceptTcpClientAsync(_tcpListener, cancellationToken)
...
         {
-            try
-            {
-                return await listener.AcceptTcpClientAsync().ConfigureAwait(false);
-            }
-            catch (Exception ex) when (token.IsCancellationRequested)
+            using (ct.Register(listener.Stop))
             {
-                throw new OperationCanceledException("Cancellation was requested while awaiting TCP client connection.", ex);
+                try
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.net45/Infrastructure/SingleMessageTcpListener.cs]: I will move this logic into the product SingleMessageTcpListener.cs or consolidate the 2 classes.

File: tests/Microsoft.Identity.Test.Integration.net45/SeleniumTests/InteractiveFlowTests.cs [hunk 3]
```diff
@@ -216,124 +213,124 @@ public async Task Interactive_AdfsV2019_DirectAsync()
...
+
+
         [TestMethod]
-        [Ignore("Lab needs a way to provide multiple account types(AAD, ADFS, MSA) that can sign into the same client id")]
-        public async Task MultiUserCacheCompatabilityTestAsync()
-        {
-            // Arrange
-            //cache = new TokenCache();
+        [Ignore("Lab needs a way to provide multiple account types(AAD, ADFS, MSA) that can sign into the same client id")]
+        public async Task MultiUserCacheCompatabilityTestAsync()
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.net45/SeleniumTests/InteractiveFlowTests.cs]: Most of these changes are whitespace related :( More VS shenanigans

File: tests/Microsoft.Identity.Test.Integration.net45/SeleniumTests/SeleniumInfrastructureTests.cs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            TestCommon.ResetInternalStaticCaches();
+        }
+
+        #endregion
+
+        [TestMethod]
+        public async Task FailingTest_SeleniumFailureAsync()
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.net45/SeleniumTests/SeleniumInfrastructureTests.cs]: Tests for test infrastructure :)

**Comment by bgavrilMS** [comment]
Passing tests: https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=345582

File: tests/Microsoft.Identity.Test.Integration.net45/SeleniumTests/InteractiveFlowTests.cs [hunk 5]
```diff
@@ -216,124 +191,150 @@ public async Task Interactive_AdfsV2019_DirectAsync()
 
             LabResponse labResponse = LabUserHelper.GetLabUserData(query);
             await RunTestForUserAsync(labResponse, true).ConfigureAwait(false);
-        }
-
+        }
+
+
```
Comment (trwalke) [inline tests/Microsoft.Identity.Test.Integration.net45/SeleniumTests/InteractiveFlowTests.cs]: extra line here

File: tests/Microsoft.Identity.Test.Integration.net45/SeleniumTests/InteractiveFlowTests.cs [hunk 6]
```diff
@@ -216,124 +191,150 @@ public async Task Interactive_AdfsV2019_DirectAsync()
...
-        	return result;
-		}
 
-        private SeleniumWebUI CreateSeleniumCustomWebUI(LabUser user, Prompt prompt, bool withLoginHint, bool adfsOnly = false)
-        {
-            return new SeleniumWebUI((driver) =>
-            {
+        }
+
+
```
Comment (trwalke) [inline tests/Microsoft.Identity.Test.Integration.net45/SeleniumTests/InteractiveFlowTests.cs]: nit: extra line here

File: tests/Microsoft.Identity.Test.Integration.net45/SeleniumTests/SeleniumInfrastructureTests.cs (hunk 4 above)
Comment (trwalke) [reply tests/Microsoft.Identity.Test.Integration.net45/SeleniumTests/SeleniumInfrastructureTests.cs]: Good idea!

---
In reply to: [287452257](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1172#discussion_r287452257) [](ancestors = 287452257)

**Pull request merged**